### PR TITLE
feat: add all required endpoints

### LINKED ISSUE #5: URI schema definition
API endpoint details / schema

### LINKED ISSUE #6: Implement endpoints with FastAPI
Implement the endpoints using FastAPI + Boto3 + SQLAlchemy

### LINKED ISSUE #7: Define Error Handlers - 40x HTTP responses
Handle different error scenarios and return proper error HTTP response status code

### LINKED ISSUE #8: Setup cache to store Athena query IDs

### LINKED ISSUE #12: Add Dockerfile

### LINKED ISSUE #13: Add logging

### LINKED ISSUE #14: Create new endpoint for result preview
Result preview is part of status ednpoints, extract it out in a different endpoint

### LINKED ISSUE #15: Convert request query endpoint to JSON HAL spec.
https://tools.ietf.org/id/draft-kelly-json-hal-01.html

Convert the query endpoint to application/hal+json type to add hyperlinks for status and preview endpoints

### PR DESCRIPTION
**Pull request by rohitxsh**

### CODE REVIEW

**Review by andrewyatz — APPROVED** [review]
As a first pass this looks good to me

File: main.py [hunk 1]
```diff
@@ -1,15 +1,77 @@
...
 app = FastAPI()
+r = redis.Redis(host='localhost', port=6379, db=0)
+athena_client = boto3.client('athena')
+s3_client = boto3.client('s3')
+
+AWS_DATA_CATALOG = "AwsDataCatalog"
+AWS_SCHEMA_DATABASE_NAME = "ensembl-parquet-meta-schema"
+AWS_S3_OUTPUT_DIR = "s3://ensembl-athena-results/"
+
+@app.get("/api/v1")
```
Comment (marcoooo) [inline main.py]: You'd better think of your app as "behind a proxy" e.g. the relative uri is never mentioned in your code, so the day you switch to /api/v2 your routes definition don't change. More info here: https://fastapi.tiangolo.com/advanced/behind-a-proxy/
Therefore your route must be defined as "relative", this one should ten become
```suggestion
@app.get("/")
```

File: main.py [hunk 2]
```diff
@@ -1,15 +1,77 @@
...
+def root():
+    return "Ensembl's data lakehouse backend"
+
+
+@app.get("/api/v1/data")
```
Comment (marcoooo) [inline main.py]: Following first comment this one should become
```suggestion
@app.get("/data")
```

File: main.py [hunk 3]
```diff
@@ -1,15 +1,77 @@
...
+    # remove unnecessary data from AWS response
+    for table in query_response:
+        del table["CreateTime"]
+        del table["LastAccessTime"]
+        del table["TableType"]
+        del table["Parameters"]
+    return query_response
+
+
+@app.get("/api/v1/{data}/columns")
```
Comment (marcoooo) [inline main.py]: Same comment, and I would ass that {data} could be misleading with previous route, if it's a `data_type` then change the name for this. 
```suggestion
@app.get("/{data_type}/columns")
```

File: main.py [hunk 4]
```diff
@@ -1,15 +1,77 @@
...
+    conn = connect(s3_staging_dir=AWS_S3_OUTPUT_DIR, schema_name=AWS_SCHEMA_DATABASE_NAME)
+    species = pd.read_sql_query(f"SELECT DISTINCT species from {data}", conn)["species"].tolist()
+    return {'columns': athena_client.get_table_metadata(CatalogName=AWS_DATA_CATALOG, DatabaseName=AWS_SCHEMA_DATABASE_NAME, TableName=data)["TableMetadata"]["Columns"],
+            'species': species}
 
 
-@app.get("/")
-def read_root():
-    return {"Hello": "World"}
+@app.get("/api/v1/query/{queryId}/status")
```
Comment (marcoooo) [inline main.py]:
```suggestion
@app.get("/query/{queryId}/status")
```

File: main.py [hunk 5]
```diff
@@ -1,15 +1,77 @@
...
+        if "not yet finished" in str(err):
+            return {'status': "IN PROGRESS"}
+        print(err)
+        return {'status': "Encountered an error, try again!"}
 
 
-@app.get("/items/{item_id}")
-def read_item(item_id: int, q: Union[str, None] = None):
-    return {"item_id": item_id, "q": q}
+@app.get("/api/v1/query/{data}/{species}")
```
Comment (marcoooo) [inline main.py]:
```suggestion
@app.get("/query/{data_type}/{species}")
```

**Review by marcoooo — CHANGES_REQUESTED** [review]

**Review by marcoooo — APPROVED** [review]

**Review by bilalebi — APPROVED** [review]

**Comment by rohitxsh** [comment]
Changes:
Have split the endpoint for status and preview
Update query endpoint as per content-type content-type: application/hal+json (https://tools.ietf.org/id/draft-kelly-json-hal-01.html)
Preview endpoint now only return 25 rows excluding header
Sort the query string to generate base64 which is used as cache key for redis
Now caching /data and {data_type/columns too
Query endpoint now accepts custom fields too as comma separated values
Redis host and port can be updated by setting env. variable, details added in readme
Added dockerfile, details added in readme
Moved the code to a module
Added logging, added a middleware which logs details like url path, time taken, status code
Added new header X-Correlation-ID which can be used to search for error in logs
Customised the doc generated by FastAPI which can now be used a full fledged documentation for the API endpoints

**Review by andrewyatz — APPROVED** [review]

**Pull request merged**